Makefile

```diff
@@ -1,86 +1,16 @@
-BUILD=_build
+build:
+	dune build
 
-# By default (if left blank), the generator script searches for
-# the opencv4 headers in /usr/include/ and /usr/local/include/.
-# Set this variable if your opencv4 headers are located elsewhere.
-SYSTEM_INCLUDE_DIR=
+install:
+	dune install
 
-GEN=gen
-GENERATOR=generator.py
-GENERATOR_SRC=$(GENERATOR) hdr_parser.py type_manager.py incl/*
+uninstall:
+	dune uninstall
 
-CPP=$(GEN)/opencv.cpp glue.cpp
-HEADERS=$(GEN)/opencv.h glue.h
-SHARED_LIB=libocamlopencv.so
-SHARED_LIBS_INSTALL_DIR=/usr/lib/
-INSTALLED_SHARED_LIB=$(SHARED_LIBS_INSTALL_DIR)/$(SHARED_LIB)
+run:
+	dune run
 
-MLS=$(GEN)/opencv
-FLAGS=-use-ocamlfind
+doc:
+	dune build @doc
 
-LIB_NAME=opencv
-LIB_FILES=META $(BUILD)/$(GEN)/opencv.cmxa $(BUILD)/$(GEN)/*.cmx \
-	$(BUILD)/$(GEN)/*.cmi $(BUILD)/$(GEN)/*.a $(BUILD)/$(GEN)/*.cma
-NATIVE_LIB=$(GEN)/opencv.cmxa
-BYTE_LIB=$(GEN)/opencv.cma
-BUILT_LIBS=$(BUILD)/$(NATIVE_LIB) $(BUILD)/$(BYTE_LIB)
-
-DEMOS=$(wildcard demos/*)
-DEMOS_CLEAN=$(DEMOS:=.clean)
-
-default: install
-
-$(GEN)/%: ${GENERATOR_SRC}
-	@echo "Running generator script"
-	mkdir -p $(GEN)
-	./$(GENERATOR) $(GEN) $(SYSTEM_INCLUDE_DIR)
-
-$(BUILD)/$(SHARED_LIB): $(CPP) $(HEADERS)
-	@echo "Compiling shared library"
-	mkdir -p $(BUILD)
-	g++ --shared -fPIC -g -std=c++11 -pedantic -Werror -Wall $(CPP) \
-		-I `ocamlc -where` `pkg-config --libs --cflags opencv4` \
-		-o $(BUILD)/$(SHARED_LIB)
-	chmod 755 $(BUILD)/$(SHARED_LIB)
-
-$(BUILT_LIBS): $(BUILD)/$(SHARED_LIB) $(MLS:=.ml) $(MLS:=.mli)
-	@echo "Building opencv OCaml library"
-	ocamlbuild $(FLAGS) $(NATIVE_LIB) $(BYTE_LIB)
-
-$(INSTALLED_SHARED_LIB): $(BUILD)/$(SHARED_LIB)
-	@echo "Installing shared library"
-#	needing sudo is unfortunate but we need to put the shared library
-#	somewhere that the linker will find it
-	sudo rm -f $(INSTALLED_SHARED_LIB)
-	sudo cp $(BUILD)/$(SHARED_LIB) $(INSTALLED_SHARED_LIB)
-
-libinstall: $(BUILT_LIBS)
-	@echo "Installing ocamlfind library"
-	ocamlfind remove $(LIB_NAME)
-	ocamlfind install $(LIB_NAME) $(LIB_FILES)
-
-sharedlib: $(BUILD)/$(SHARED_LIB)
-lib: $(BUILT_LIBS)
-install: $(INSTALLED_SHARED_LIB) libinstall
-
-docs: $(BUILT_LIBS)
-	ocamlbuild $(FLAGS) -docflags -stars opencv.docdir/index.html
-
-$(DEMOS): $(INSTALLED_SHARED_LIB) libinstall
-	@echo "Running demo: $@"
-#	clean first to guarantee that we pick up changes to the library
-	$(MAKE) -C $@ clean run
-
-$(DEMOS_CLEAN):
-#	this is a stupid hack
-	$(MAKE) -C $(basename $@) clean
-
-clean: $(DEMOS_CLEAN)
-	ocamlbuild -clean
-	rm -f $(BUILD)/$(SHARED_LIB)
-	rm -rf $(GEN)
-	sudo rm -f $(INSTALLED_SHARED_LIB)
-	ocamlfind remove $(LIB_NAME)
-
-.PHONY: default sharedlib lib libinstall install docs clean \
-	$(DEMOS) $(DEMOS_CLEAN)
+.PHONY: build install uninstall run doc
```

Comment:
I'd suggest having a `.PHONY` declaration before each rule, instead of this all-encompassing one.

Reply:
Yes. I was following the pattern already there, so Ill leave it for now.